pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,172 @@
+
+	rawPatterns = append(rawPatterns, p...)
+
+	switch {
+	case os.IsNotExist(err):
+		return &defaultIgnorer{}, nil
+	case err != nil:
+		return nil, err
+	}
+
+	for _, ip := range p {
+		x, err := toFileIgnorePattern(ip)
+		if err != nil {
+			return nil, err
+		}
+		ignorePatterns = append(ignorePatterns, *x)
+	}
+
+	return &dockerIgnorer{}, nil
+}
+
+// scanAndBuildPatternsList takes the file typically the .dockerignore file
+// splits the file by new line (\n) and normalize them with following rules
+// - removes the UTF8 Byte Order Mark (BOM) characters
+// - scanner comments line starting with "#"
+// - trim spaces in the pattern
+// - make the pattern as clean filenames using golang filepath utils
+func scanAndBuildPatternsList(ignoreFile string) ([]string, error) {
+
+	var patterns []string
+	patterns = append(patterns, defaultPatterns...)
+
+	fr, err := os.Open(ignoreFile)
+
+	switch {
+	case os.IsNotExist(err):
+		return patterns, nil
+	case err != nil:
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(fr)
+
+	// UTF8 byte order mark (BOM) which are typically first three bytes of the file with
+	// the hexadecimal characters: EF,BB,BF
+	utf8bom := []byte{0xEF, 0xBB, 0xBF}
+	lineNo := 0
+
+	for scanner.Scan() {
+		scannedBytes := scanner.Bytes()
+
+		if lineNo == 0 {
+			scannedBytes = bytes.TrimPrefix(scannedBytes, utf8bom)
+		}
+
+		pattern := string(scannedBytes)
+		// removes the all unwanted spaces from the beginning and end of the string
+		// e.g. "\n\t\nfoo bar\n\t" will be trimmed to "foo bar"
+		pattern = strings.TrimSpace(pattern)
+		lineNo++
+
+		// When the line starts with comments like #, scanner the line.
+		// e.g. dockerignore
+		//   # comment to scanner target
+		//   target
+		if strings.HasPrefix("#", pattern) {
+			continue
+		}
+
+		// Skip Empty lines that might be present in the file
+		if pattern == "" {
+			continue
+		}
+
+		// The Patterns can start with ! symbolizing inversion of the pattern
+		// When ! is seen remove the ! to clean up the pattern for Path separators
+		// e.g. Patterns line !README.md
+		invert := strings.HasPrefix(pattern, invertPrefix)
+		if invert {
+			pattern = strings.TrimPrefix(pattern, invertPrefix)
+		}
+
+		if len(pattern) > 0 {
+			pattern = filepath.Clean(pattern)
+			pattern = filepath.ToSlash(pattern)
+			if len(pattern) > 1 && strings.HasPrefix(pattern, string(os.PathSeparator)) {
+				pattern = strings.TrimPrefix(pattern, string(os.PathSeparator))
+			}
+		}
+		// prefix the clean pattern with invertPrefix
+		if invert {
+			pattern = invertPrefix + pattern
+		}
+
+		patterns = append(patterns, pattern)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error reading file %s : %w", ignoreFile, err)
+	}
+	return patterns, nil
+}
+
+// Scan is the directory scanner to scan directory for .dockerignore based
+// ignore patterns
+func (id *dockerIgnorer) Scan() ([]string, error) {
+	// TODO (kamesh,matt) - move this to use godirwalk
```

Comment:
I may try to knock this off tonight, so I can measure how much faster it is

Reply:
based on some simple local measurements (macOS) this doesn't seem to be any faster than `filepath.Walk`.

median timings of 10x bundling `knative.dev/serving`:
`filepath.Walk`: 3.26s
`godirwalk` (sorted): 3.51s
`godirwalk` (unsorted): 3.38s

Reply:
Ok then, lets not worry about that now.

Reply:
removed those TODO for now and lets come back to it later